Khan/uglifyjs-webpack-plugin
# feat: don't compute input source maps if source file hasn't changed

This PR contains a:

- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

The plugin does a lot of work computing input source maps when it isn't necessary, i.e. the cache key hasn't changed and we already have a source map in our cache for the key.  For our particular use case I saw a reduction in fully cached build times from UglifyJSPlugin taking 90s to 10s of the build time.

### Breaking Changes

This shouldn't be a breaking change, but I'm not super familiar with all of the possible configurations for this plugin.

### Additional Info

This change looks a lot bigger than it actually is.  Because of the `cacache.get.info()` call that's introduced index, I had to wrap a big chunk of `optimizeFn` in a `Promise.all()` call because `cacache.get.info()` returns a promise and before this everything in `optimizeFn` was synchronous.

NOTE: I've also include the build files in `dist` so that we can `npm install` the module from this commit without having to do a publish.

I've benchmarked doing a fresh build and a repeat build both with and without these changes in webapp.  Here are the results:
```
fresh build without changes: 
430150ms chunk asset optimization (uglifyjs) = 8m 10s
real    10m20.638s
user    26m1.687s
sys    2m25.869s

repeat build without changes:
154745ms chunk asset optimization (uglifyjs) = 2m 35s
real    3m20.630s
user    3m42.489s
sys    0m21.300s

fresh build with changes: 
437674ms chunk asset optimization (uglifyjs) = 7m 18s
real    10m9.119s
user    26m1.528s
sys    2m32.005s

repeat build with changes: 
7786ms chunk asset optimization (uglifyjs) = 0m 8s
real    0m45.423s
user    0m46.280s
sys    0m6.520s
```

### Test Plan

- run `yarn build:webpack-prod-profile --no-i18n` in webapp
- see `10535ms chunk asset optimization` which represents UglifyJSPlugin's time (this was close to 90 seconds before this change)
- see `12692ms after chunk asset optimization` which represents SourceMapDevToolPlugin's time (no change to this time, only included here to indicate that UglifyJSPlugin is no longer the bottleneck)
- run `md5` on some of the files in `genwebpack/prod/en/javascript` and see that they match the values before the changes.
- run `npm test`
- checksum all of the files output by webpack using `find genwebpack/prod/en -type f -exec shasum "{}" ';' > checksumsN.txt` and `diff checksumA.txt checksumB.txt` to compare them and see that they all match up.

<img width="781" alt="screen shot 2019-01-16 at 11 14 41 am" src="https://user-images.githubusercontent.com/1044413/51262226-f3014b00-197f-11e9-8173-b8351f0c1cd7.png">
